### ISSUE TITLE
[RFC/Runtime] Missing bulk buffer/slice initialization and copy primitives (buffer.zeroed, buffer.fill, slice.copy_from, slice.fill)

### ISSUE THREAD

**Issue by sanohiro** [sanohiro/align#1051]
# [RFC/Runtime] Missing bulk buffer/slice initialization and copy primitives (`buffer.zeroed`, `buffer.fill`, `slice.copy_from`, `slice.fill`)

## Summary

In `align-llm`, buffers frequently need bulk allocation and initialization with uniform values (e.g. zeroed vocabulary logits for 151,936 tokens = 607 KB, attention masks filled with `-inf` bits = 128 KB, zeroed KV slot tables = 64 KB).

Currently, Align's `buffer` and `slice<u8>` standard library lacks bulk initialization and copy operations:
1. `buffer(capacity)` only reserves capacity; `b.len` is 0.
2. The only standard way to fill a buffer is appending element-by-element (`b.put_u32_le(0)` or `b.put_u8(0)`), requiring $O(N)$ runtime FFI calls and vector resizes (e.g. 151,936 iterations for logits).
3. Attempting an exponential doubling pattern in userland (`b.append(b.bytes()[0..step])`) triggers buffer address-overlap detection in `align_rt_buffer_append` (`crates/align_runtime/src/lib.rs:11532`), forcing a temporary heap `Vec` allocation (`to_vec()`) on every doubling step.
4. `slice<u8>` lacks in-place `copy_from` (mapping to `memcpy` / `copy_from_slice`) and `fill` (mapping to `memset`).

---

## 1. Concrete Problem & Cost in `align-llm`

In `src/runtime_generation.align`:
- `logits_buffer(n_vocab: 151936)`: Needs 607,744 zero bytes.
- `mask_storage`: Needs 131,072 bytes filled with `0xFF800000` (`-inf`).
- `Session.empty`: Needs 65,536 zero bytes for KV cache prefix initialization.

#### Approach A: Element-by-element loop
```align
mut at := 0
loop {
  if at >= n_vocab { break }
  b.put_u32_le(0)
  at = at + 1
}
```
- **Cost**: 151,936 iterations in userland. Each call invokes `align_rt_buffer_put` across the FFI boundary, truncating and extending the buffer `Vec`. On CPU this costs noticeable latency on session / prompt creation.

#### Approach B: Userland exponential doubling
```align
pub fn filled_buffer_u32(count: i64, value: u32) -> Result<buffer, Error> {
  ...
  loop {
    if b.bytes().len() >= bytes { break }
    remaining := bytes - b.bytes().len()
    current_len := b.bytes().len()
    step := if remaining < current_len { remaining } else { current_len }
    b.append(b.bytes()[0..step])
  }
  ...
}
```
- **Limitation**: In `crates/align_runtime/src/lib.rs:11532`:
  ```rust
  let aliases_buffer = matches!(
      (src_end, buf_end),
      (Some(se), Some(be)) if src_start < be && buf_start < se
  );
  let snapshot =
      aliases_buffer.then(|| unsafe { core::slice::from_raw_parts(ptr, n) }.to_vec());
  ```
  Because the source slice aliases `b.data`, `align_rt_buffer_append` detects the alias and allocates a temporary `Vec` for every step. While reducing iteration count to $\approx 12$, it incurs repeated temporary heap allocations.

---

### 1. `buffer.zeroed(len: i64) -> Result<buffer, Error>`
Creates a pre-sized buffer of length `len` initialized to zero (equivalent to `vec![0u8; len]` / `calloc`).
- Instant $O(1)$ page-table / zero-page mapping for large allocations, or single `memset`.
- Returns a buffer with `len == capacity == len`.

### 2. `buffer.filled(count: i64, byte: u8) -> Result<buffer, Error>`
Creates a pre-sized buffer of length `count` initialized with `byte` via `memset`.

### 3. `slice<u8>.fill(byte: u8)`
Fills an existing mutable byte slice in-place via libc `memset`.

### 4. `slice<u8>.copy_from(borrow src: slice<u8>) -> Result<(), Error>`
Copies bytes from `src` into `dest` in-place via libc `memcpy` / `memmove`.
- Validates lengths (`src.len() == dest.len()` or `src.len() <= dest.len()`).
- Emits optimized LLVM `llvm.memcpy` intrinsic (hardware SIMD accelerated).

---

## 3. Benefits to Real Clients
- Eliminates 100% of userland loop overhead and FFI crossing overhead for buffer initialization.
- Eliminates temporary heap snapshots during exponential doubling.
- Provides standard, idiomatic, memory-bandwidth-speed bulk data initialization for machine learning and system applications.

**Comment by sanohiro** [comment]
The initialized-allocation part is merged in #1056 as `buffer.filled(length: i64, value: u8) -> buffer`. It guarantees initialized length without a userland append loop. It follows core terminal allocation failure, not a new Result allocator; capacity may exceed length, and no constant-time zero-page guarantee is made.

The remaining in-place byte fill/copy and repeated-word fill family are accepted in plan 65 and remain under implementation with the complete access/ownership capability in plan 61. They are not shipped yet. In particular, a byte fill is not a substitute for repeating the f32 negative-infinity bit pattern. Keeping this issue open for those remaining surfaces.

**Comment by sanohiro** [comment]
In-place byte fill/copy and pattern fill implemented in branch `feat/array-truncate` (commit `eca0cf62`).

- `slice<u8>.fill(value: u8) -> ()` (libc memset lowering)
- `slice<u8>.fill_S_E(value: S) -> ()` for all 8 multi-byte scalars and endian suffixes (divisible length check)
- `slice<u8>.copy_from(source: slice<u8>) -> ()` with compile-time backing storage non-overlap proof and equal-length runtime validation (LLVM memcpy lowering)

All mutation receivers enforce mutable/writable places; overlapping/unknown backing buffers are rejected. Verified across focused owner tests (`tests/bytes_ops.rs`), bounded gate, and workspace Clippy. Ready for PR.
